// tokenProfiles/googleIdToken.js
'use strict';

var jwt = require('jsonwebtoken');
var BasicIdToken = require('./basicIdToken');  


function GoogleIdToken(name, email, picture, iss, sub, aud, iat, exp){
    this.name = name;
    this.email = email;
    this.picture = picture;
    this.iss = iss;
    this.sub = sub;
    this.aud = aud;
    this.iat = iat;
    this.exp = exp;
    this.validateRequiredFields();
};

GoogleIdToken.prototype = Object.create(BasicIdToken.prototype);
GoogleIdToken.prototype.constructor = ExtendedIdToken;

GoogleIdToken.prototype.options_to_payload = {
    'name': 'name',
    'email': 'email',
    'picture': 'picture',
    'iss': 'iss',
    'sub': 'sub',
    'aud': 'aud',
    'iat': 'iat',
    'exp': 'exp',
};
  
GoogleIdToken.prototype.options_for_objects = [
    'expiresIn',
    'notBefore',
    'noTimestamp',
    'audience',
    'issuer',
    'subject',
    'jwtid',
];

GoogleIdToken.prototype.knownNonStandardClaims = {
    'jti' : 'jti',
};

GoogleIdToken.prototype.validateRequiredFields = function(){
    if (this.name && this.email && this.picture && this.iss && this.sub && this.aud && this.iat && this.exp){
        console.log("Validated all standard fields")
    }else {
        throw new Error("You are missing a required parameter");
    }
};

GoogleIdToken.prototype.getStandardClaims = function(){
    GoogleIdToken.prototype.standard_claims = {"name": this.name, "email" : this.email, "picture": this.picture,  "iss" : this.iss, "sub" : this.sub, "aud" : this.aud, "iat": this.iat, "exp": this.exp};
    return GoogleIdToken.prototype.standard_claims;         
};

module.exports = ExtendedIdToken;
