// tokenProfiles/implicitAccessToken.js
'use strict';

var jwt = require('jsonwebtoken');
var BasicIdToken = require('./basicIdToken');  

function ImplicitAccessToken(iss, sub, aud, iat){
    this.iss = iss;
    this.sub = sub;
    this.aud = aud;
    this.iat = iat;
    this.validateRequiredFields();
};

ImplicitAccessToken.prototype = Object.create(BasicIdToken.prototype);
ImplicitAccessToken.prototype.constructor = ImplicitAccessToken;

ImplicitAccessToken.prototype.options_to_payload = {
    'iss': 'iss',
    'sub': 'sub',
    'aud': 'aud',
    'iat': 'iat',
};
  
ImplicitAccessToken.prototype.options_for_objects = [
    'expiresIn',
    'notBefore',
    'noTimestamp',
    'audience',
    'issuer',
    'subject',
    'jwtid',
];

ImplicitAccessToken.prototype.knownNonStandardClaims = {
    'knownNonStandardClaim' : 'knownNonStandardClaim',
};

ImplicitAccessToken.prototype.validateRequiredFields = function(){
    if (this.iss && this.sub && this.aud && this.iat){
        console.log("Validated all standard fields")
    }else {
        throw new Error("You are missing a required parameter");
    }
};

ImplicitAccessToken.prototype.getStandardClaims = function(){
    ImplicitAccessToken.prototype.standard_claims = { "iss" : this.iss, "sub" : this.sub, "aud" : this.aud, "iat": this.iat};
    return ImplicitAccessToken.prototype.standard_claims;         
};

module.exports = ImplicitAccessToken;
