// tokenProfiles/scopedAccessToken.js
'use strict';

var jwt = require('jsonwebtoken');
var BasicIdToken = require('./basicIdToken');  

function ScopedAccessToken(iss, sub, aud, iat, exp, scope){
    this.iss = iss;
    this.sub = sub;
    this.aud = aud;
    this.iat = iat;
    this.exp = exp;
    this.scope = scope;
    this.validateRequiredFields();
};

ScopedAccessToken.prototype = Object.create(BasicIdToken.prototype);
ScopedAccessToken.prototype.constructor = AccessToken;

ScopedAccessToken.prototype.options_to_payload = {
    'iss': 'iss',
    'sub': 'sub',
    'aud': 'aud',
    'iat': 'iat',
    'exp': 'exp',
    'scope': 'scope',
};
  
ScopedAccessToken.prototype.options_for_objects = [
    'expiresIn',
    'notBefore',
    'noTimestamp',
    'audience',
    'issuer',
    'subject',
    'jwtid',
];

ScopedAccessToken.prototype.knownNonStandardClaims = {
    'knownNonStandardClaim' : 'knownNonStandardClaim',
};

ScopedAccessToken.prototype.validateRequiredFields = function(){
    if (this.iss && this.sub && this.aud && this.iat && this.exp && this.scope){
        console.log("Validated all standard fields")
    }else {
        throw new Error("You are missing a required parameter");
    }
};

ScopedAccessToken.prototype.getStandardClaims = function(){
    ScopedAccessToken.prototype.standard_claims = { "iss" : this.iss, "sub" : this.sub, "aud" : this.aud, "iat": this.iat, "exp": this.exp, "scope" : this.scope};
    return ScopedAccessToken.prototype.standard_claims;         
};

module.exports = ScopedAccessToken;
