// tokenProfiles/accessToken.js
'use strict';

var jwt = require('jsonwebtoken');
var BasicIdToken = require('./basicIdToken');  

function AccessToken(iss, sub, iat){
    this.iss = iss;
    this.sub = sub;
    this.iat = iat;
    this.validateRequiredFields();
};

AccessToken.prototype = Object.create(BasicIdToken.prototype);
AccessToken.prototype.constructor = AccessToken;

AccessToken.prototype.options_to_payload = {
    'iss': 'iss',
    'sub': 'sub',
    'iat': 'iat',
};
  
AccessToken.prototype.options_for_objects = [
    'expiresIn',
    'notBefore',
    'noTimestamp',
    'audience',
    'issuer',
    'subject',
    'jwtid',
];

AccessToken.prototype.knownNonStandardClaims = {
    'aud': 'aud',
    'exp': 'exp',
};

AccessToken.prototype.validateRequiredFields = function(){
    if (this.iss && this.sub && this.iat){
        console.log("Validated all standard fields")
    }else {
        throw new Error("You are missing a required parameter");
    }
};

AccessToken.prototype.getStandardClaims = function(){
    AccessToken.prototype.standard_claims = { "iss" : this.iss, "sub" : this.sub, "iat": this.iat};
    return AccessToken.prototype.standard_claims;         
};

module.exports = AccessToken;
