// tokenProfiles/implicitAccessToken.js
'use strict';

var jwt = require('jsonwebtoken');
var BasicIdToken = require('./basicIdToken');  

function ImplicitAccessToken(iss, sub, iat){
    this.iss = iss;
    this.sub = sub;
    this.iat = iat;
    this.validateRequiredFields();
};

ImplicitAccessToken.prototype = Object.create(BasicIdToken.prototype);
ImplicitAccessToken.prototype.constructor = ImplicitAccessToken;

ImplicitAccessToken.prototype.options_to_payload = {
    'iss': 'iss',
    'sub': 'sub',
    'iat': 'iat',
};
  
ImplicitAccessToken.prototype.options_for_objects = [
    'expiresIn',
    'notBefore',
    'noTimestamp',
    'audience',
    'issuer',
    'subject',
    'jwtid',
];

ImplicitAccessToken.prototype.knownNonStandardClaims = {
    'aud' : 'aud',
};

ImplicitAccessToken.prototype.validateRequiredFields = function(){
    if (this.iss && this.sub && this.iat){
        console.log("Validated all standard fields")
    }else {
        throw new Error("You are missing a required parameter");
    }
};

ImplicitAccessToken.prototype.getStandardClaims = function(){
    ImplicitAccessToken.prototype.standard_claims = { "iss" : this.iss, "sub" : this.sub, "iat": this.iat};
    return ImplicitAccessToken.prototype.standard_claims;         
};

ImplicitAccessToken.prototype.claims_to_verify = {
    'iss': 'iss',
    'sub': 'sub',
    'maxAge' : 'maxAge',
};

ImplicitAccessToken.prototype.initData = function(){
    ImplicitAccessToken.prototype.non_standard_verification_claims = {};    
    ImplicitAccessToken.prototype.NoneAlgorithm = false;
};


ImplicitAccessToken.prototype.addNonStandardClaims = function(nonStandardClaims){
    ImplicitAccessToken.prototype.non_standard_claims = nonStandardClaims;

    ImplicitAccessToken.prototype.non_standard_verification_claims = {};
    Object.keys(nonStandardClaims).forEach(function (key) {
        if (ImplicitAccessToken.prototype.knownNonStandardClaims[key]) {
            ImplicitAccessToken.prototype.non_standard_verification_claims[key] = nonStandardClaims[key];
        }
    });  
};

ImplicitAccessToken.prototype.getNonStandardClaims = function(nonStandardClaims){
    return ImplicitAccessToken.prototype.non_standard_claims;
}; 

ImplicitAccessToken.prototype.getVerificationClaims = function(){
    return ImplicitAccessToken.prototype.verification_claims;
}; 

ImplicitAccessToken.prototype.getNonStandardVerificationClaims = function(){
    return ImplicitAccessToken.prototype.non_standard_verification_claims;
}; 

/* User explicitly wants to set None Algorithm attribute */
ImplicitAccessToken.prototype.setNoneAlgorithm = function(boolVal){
    ImplicitAccessToken.prototype.NoneAlgorithm = boolVal;
};

ImplicitAccessToken.prototype.getNoneAlgorithm = function(boolVal){
    return ImplicitAccessToken.prototype.NoneAlgorithm;
};

/* Support for different serialization and deserialiation types */
ImplicitAccessToken.prototype.toJWT = function(secretOrPrivateKey, options){
    return jwt.sign(this, secretOrPrivateKey, options);
};

ImplicitAccessToken.prototype.fromJWT = function(signedJWT, secretOrPrivateKey, claimsToVerify, options){

    this.validateRequiredVerificationClaims(claimsToVerify);
    this.validateRequiredNonStandardVerificationClaims(claimsToVerify);
    return jwt.decode(signedJWT,secretOrPrivateKey, this, options);
};

ImplicitAccessToken.prototype.validateRequiredVerificationClaims = function(claimsToVerify)
{
    Object.keys(ImplicitAccessToken.prototype.claims_to_verify).forEach(function (key) {
        if (!claimsToVerify[key]) {
            throw new Error('Missing required verification claim: ' + key);
        }
      });  
      ImplicitAccessToken.prototype.verification_claims = claimsToVerify;
};

ImplicitAccessToken.prototype.validateRequiredNonStandardVerificationClaims = function(claimsToVerify)
{
    if (ImplicitAccessToken.prototype.non_standard_verification_claims['aud']){
        this.nonStandardVerificationClaimsCheck('aud', claimsToVerify);
    }
};

ImplicitAccessToken.prototype.nonStandardVerificationClaimsCheck = function(key, claimsToVerify){
    if (!claimsToVerify[key]) {
        throw new Error('Missing required verification claim: ' + key);
    }else{
        ImplicitAccessToken.prototype.verification_claims[key] = claimsToVerify[key];
        if (key == "aud"){
            ImplicitAccessToken.prototype.claims_to_verify['aud'] = 'aud';
        }
    }
}

module.exports = ImplicitAccessToken;
