// tokenProfiles/scopedAccessToken.js
'use strict';

var jwt = require('jsonwebtoken');
var AccessToken = require('./accessToken');  

function ScopedAccessToken(iss, sub, iat, scope){
    this.iss = iss;
    this.sub = sub;
    this.iat = iat;
    this.scope = scope;
    this.validateRequiredFields();
};

ScopedAccessToken.prototype = Object.create(AccessToken.prototype);
ScopedAccessToken.prototype.constructor = AccessToken;

ScopedAccessToken.prototype.options_to_payload = {
    'iss': 'iss',
    'sub': 'sub',
    'iat': 'iat',
    'scope': 'scope',
};
  
ScopedAccessToken.prototype.options_for_objects = [
    'expiresIn',
    'notBefore',
    'noTimestamp',
    'audience',
    'issuer',
    'subject',
    'jwtid',
];

ScopedAccessToken.prototype.knownNonStandardClaims = {
    'aud' : 'aud',
    'exp' : 'exp',
};

ScopedAccessToken.prototype.validateRequiredFields = function(){
    if (this.iss && this.sub && this.iat && this.scope){
        console.log("Validated all standard fields")
    }else {
        throw new Error("You are missing a required parameter");
    }
};

ScopedAccessToken.prototype.getStandardClaims = function(){
    ScopedAccessToken.prototype.standard_claims = { "iss" : this.iss, "sub" : this.sub, "iat": this.iat, "scope" : this.scope};
    return ScopedAccessToken.prototype.standard_claims;         
};

module.exports = ScopedAccessToken;
